src/core/QCheck2.ml

```diff
@@ -0,0 +1,1868 @@
+
+  let int_towards destination x =
+    Int.(number_towards ~equal ~div ~add ~sub ~of_int:Fun.id) ~destination x
+
+  let int32_towards destination x =
+    Int32.(number_towards ~equal ~div ~add ~sub ~of_int) ~destination x
+
+  let int64_towards destination x =
+    Int64.(number_towards ~equal ~div ~add ~sub ~of_int) ~destination x
+
+  (** Arbitrarily limit to 15 elements as dividing a [float] by 2 doesn't converge quickly
+      towards the destination. *)
+  let float_towards destination x =
+    Float.(number_towards ~equal ~div ~add ~sub ~of_int) ~destination x |> Seq.take 15
+
+  let int_aggressive_towards (destination : int) (n : int) : int Seq.t =
+    Seq.unfold (fun current ->
+        if current = n then None
+        else if current < n then let next = succ current in Some (next, next)
+        else let next = pred current in Some (next, next)
+      ) destination
+
+  let int_aggressive n = int_aggressive_towards 0 n
+
+end
+
+module Tree = struct
+  type 'a t = Tree of 'a * ('a t) Seq.t
+
+  let root (Tree (root, _) : 'a t) : 'a = root
+
+  let children (Tree (_, children) : 'a t) : ('a t) Seq.t = children
+
+  let rec pp ?(depth : int option) (inner_pp : Format.formatter -> 'a -> unit) (ppf : Format.formatter) (t : 'a t) : unit =
+    let open Format in
+    let Tree (x, xs) = t in
+    let wrapper_box inner =
+      pp_open_vbox ppf 2;
+      pp_print_string ppf "Tree(";
+      pp_print_break ppf 0 0;
+      inner ();
+      pp_close_box ppf ();
+      pp_print_break ppf 0 0;
+      pp_print_string ppf ")"
+    in
+    let inner () =
+      pp_open_vbox ppf 2;
+      pp_print_string ppf "Node(";
+      pp_print_break ppf 0 0;
+      inner_pp ppf x;
+      pp_close_box ppf ();
+      pp_print_break ppf 0 0;
+      pp_print_string ppf "),";
+      pp_print_break ppf 1 0;
+      pp_open_vbox ppf 2;
+      pp_print_string ppf "Shrinks(";
+
+      if Option.fold depth ~none:false ~some:(fun depth -> depth <= 0) then (
+        pp_print_string ppf "<max depth reached>";
+        pp_close_box ppf ())
+      else if Seq.is_empty xs then pp_close_box ppf ()
+      else
+        (pp_print_break ppf 0 0;
+         pp_print_list
+           ~pp_sep:(fun ppf () -> pp_print_string ppf ","; pp_print_space ppf ())
+           (pp ?depth:(Option.map pred depth) inner_pp)
+           ppf
+           (List.of_seq xs);
+         pp_close_box ppf ();
+         pp_print_break ppf 0 0);
+
+      pp_print_string ppf ")"
+
+    in
+    wrapper_box inner
+
+  let rec map (f : 'a -> 'b) (a : 'a t) : 'b t =
+    let Tree (x, xs) = a in
+    let y = f x in
+    let ys = Seq.map (fun smaller_x -> map f smaller_x) xs in
+    Tree (y, ys)
+
+  (** Note that parameter order is reversed. *)
+  let (>|=) a f = map f a
+
+  let rec ap (f : ('a -> 'b) t) (a : 'a t) : 'b t =
+    let Tree (x0, xs) = a in
+    let Tree (f0, fs) = f in
+    let y = f0 x0 in
+    let ys = Seq.append (Seq.map (fun f' -> ap f' a) fs) (Seq.map (fun x' -> ap f x') xs) in
```

Comment:
Doesn't this implementation break some applicative laws that we may want to respect? 

It breaks at least the composition property:
```ocaml
pure (fun g f x -> g (f x)) <*> u <*> v <*> w = u <*> (v <*> w)
```
but I don't know if this can lead to problems along the way, maybe we don't care

Also: functionally speaking, what is the benefit of such interleaving? 

I **think** a benefit is that if the problem is on value `a` (no matter the function `a -> b`), it will try faster shrinks of `a` (instead of first trying to exhaust all shrinks of `f`, and also trying again all shrinks of `f` at each shrink of `a` before further shrinking `a`), am I correct?

This is a sensible benefit indeed

Reply:
The (expected) benefit is indeed to be able to try shrinks of `b` before we exhaust the (direct) shrinks of `a`, so it's more "fair".

I don't think it breaks the composition property, since we don't really have a notion of equality on trees :smirk: . If you consider shrinks as a set, the associativity still holds; is order of shrinks something we should care about algebraically?